### PR TITLE
TacticalDataManager implementation!

### PR DESCRIPTION
**Pull request by alxhub**
[Review on Reviewable](https://reviewable.io/reviews/angular/tactical/24)

### CODE REVIEW

**Comment by ttowncompiled** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[modules/tactical/src/backend.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0aFVaMYipSctiDwwm)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L10)):</sup>
I believe that this should be an interface, not a class. It seems that VersionedObjects are just a type of Object. They contain no logic, only properties. Interfaces fill this role in TypeScript.

---

<sup>**[modules/tactical/src/backend.ts, line 36 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0bVLRm78-Ss7xN7aF)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L36)):</sup>
{[key: string]: VersionedObject} is the same as Array<VersionedObject> or VersionedObject[], in this case. TypeScript thinks of Arrays as objects that are indexed by both string and number and that contain no other properties. This is weird to me, but they they decided to stick with this design since JavaScript casts integer keys to string keys before resolving the index.

---

<sup>**[modules/tactical/src/data_manager.ts, line 23 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0dWVdtRHHzO1W0wqs)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L23)):</sup>
I believe this is also just Array<Link<Object>[]> or Link<Object>[][].

---

<sup>**[modules/tactical/test/link_spec.ts, line 48 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0g74wNvEPHOhBo0lj)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/test/link_spec.ts#L48)):</sup>
Clang apparently doesn't care about newlines at EOF so it's up to you if we should keep up with this.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/tactical/24)

**Comment by ttowncompiled** [comment]
LGTM pending comments.

File: modules/tactical/src/data_manager.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/// <reference path="../../../typings/rx/rx.d.ts" />
+/// <reference path="../../../typings/rx/rx-lite.d.ts" />
+
+import {Observable, Subject} from 'rx';
+import {Backend, VersionedObject} from './backend';
```
Comment (gkalpak) [inline modules/tactical/src/data_manager.ts]: `VersionedObject` doesn't seem to be used.
Not sure if it should be used or removed from the imports.

File: modules/tactical/src/link.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+   */
+  observer: any;
+
+  /**
+   * Called when all subscribers disconnect from the Observable.
+   */
+  onDisconnect: Function = () => {}
+
+  constructor() {
+    var link = this;
```
Comment (gkalpak) [inline modules/tactical/src/link.ts]: Is the alias really needed (since you are using arrow functions) ?

**Comment by ttowncompiled** [comment]
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[modules/tactical/src/data_manager.ts, line 5 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0Y1OJ-EEgIgj-JJ9t-r1-5)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L5)):</sup>
It's being used,
```typescript
_backendData(data: VersionedObject) { ...
```

---

<sup>**[modules/tactical/src/link.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0Y1OJ-EEgIgj-JJ9u-r1-29)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L29)):</sup>
It shouldn't be. In TypeScript, ```this``` inside the context of a fat arrow function is taken from the scope in which the fat arrow function is defined. I believe that you can cascade ```this``` through multiple contexts if you keep define nested fat arrow functions so ```link == this``` should be true in every instance of ```link``` in this constructor.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/tactical/24)

**Comment by ttowncompiled** [comment]
Review status: 0 of 6 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

<sup>**[modules/tactical/src/backend.ts, line 23 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4AO6DvldJhpi7nn7h)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L23)):</sup>
Angular 2 style is to add a return type of void here
```typescript
request(key: Object): void;
```

---

<sup>**[modules/tactical/src/backend.ts, line 40 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4AYedD2XdXcEzhdH6)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L40)):</sup>
Return type ```void```.

---

<sup>**[modules/tactical/src/backend.ts, line 45 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4Ai-lKIY_dOtK8pGB)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L45)):</sup>
Why not move this into the function and take away the ```!``` from the boolean expression?

---

<sup>**[modules/tactical/src/backend.ts, line 53 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4AwoLoN7l1udML9UF)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L53)):</sup>
Return type ```void```.

---

<sup>**[modules/tactical/src/data_manager.ts, line 35 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt49l7S5ofumsZLteyt)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L35)):</sup>
Why is this not private? Also, return type ```void```.

---

<sup>**[modules/tactical/src/data_manager.ts, line 40 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BHUQKZpF-dR1NZlB)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L40)):</sup>
Same here. Why not move this into the function?

---

<sup>**[modules/tactical/src/data_manager.ts, line 64 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BOll5IGPdRn5r7Iu)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L64)):</sup>
Return type ```void```.

---

<sup>**[modules/tactical/src/link.ts, line 21 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BVzOjzOGRVA81bGh)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L21)):</sup>
Would this not be Rx.Observer<T>?

---

<sup>**[modules/tactical/src/link.ts, line 39 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BdTBOyFaAC4YMjO4)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L39)):</sup>
Return type ```void```.

---

<sup>**[modules/tactical/src/link.ts, line 45 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4Bi9tenQogPFbX0j6)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L45)):</sup>
Return type ```void```.

---

<sup>**[modules/tactical/src/link.ts, line 51 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BjjGtyqHZMt1iZIS)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L51)):</sup>
Return type ```void```.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/tactical/24)

**Comment by ttowncompiled** [comment]
Review status: 0 of 6 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

<sup>**[modules/tactical/src/link.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0Y1OJ-EEgIgj-JJ9u-r1-29)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L29)):</sup>
According to the TypeScript playground, you should be fine using ```this``` instead of ```link```. TypeScript will actually define the alias for you.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/tactical/24)

**Comment by ttowncompiled** [comment]
Review status: 0 of 6 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

<sup>**[modules/tactical/src/data_manager.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4FXgpFTly4ZqpyH-P)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L28)):</sup>
This alias isn't needed.

---

<sup>**[modules/tactical/src/data_manager.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4FZmYDkP3uicxR_kr)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L29)):</sup>
s/ ```(data)``` / ```(data: VersionedObject)```

---

<sup>**[modules/tactical/src/data_manager.ts, line 40 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4FmGGSaRbOwdWI-yD)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L40)):</sup>
s/ ```(link)``` / ```(link: Link<Object>)```

---

<sup>**[modules/tactical/src/link.ts, line 30 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4GBHB9yTm9MQzem8b)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L30)):</sup>
s / ```(observer)``` / ```(observer: Observer<T>)```

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/tactical/24)

**Comment by alxhub** [comment]
Review status: 0 of 6 files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

<sup>**[modules/tactical/src/backend.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0aFVaMYipSctiDwwm)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L10)):</sup>
Done.

---

<sup>**[modules/tactical/src/backend.ts, line 23 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4AO6DvldJhpi7nn7h)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L23)):</sup>
Done.

---

<sup>**[modules/tactical/src/backend.ts, line 40 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4AYedD2XdXcEzhdH6)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L40)):</sup>
Done.

---

<sup>**[modules/tactical/src/backend.ts, line 45 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4Ai-lKIY_dOtK8pGB)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L45)):</sup>
Done (assuming you meant into the if statement).

---

<sup>**[modules/tactical/src/backend.ts, line 53 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4AwoLoN7l1udML9UF)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/backend.ts#L53)):</sup>
Done.

---

<sup>**[modules/tactical/src/data_manager.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4FXgpFTly4ZqpyH-P)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L28)):</sup>
Done.

---

<sup>**[modules/tactical/src/data_manager.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4FZmYDkP3uicxR_kr)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L29)):</sup>
Done.

---

<sup>**[modules/tactical/src/data_manager.ts, line 35 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt49l7S5ofumsZLteyt)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L35)):</sup>
Done.

---

<sup>**[modules/tactical/src/data_manager.ts, line 40 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BHUQKZpF-dR1NZlB)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L40)):</sup>
In this case, I actually prefer it as written.

In general, the **primary** flow of a function should be as high-level as possible, not buried inside error-checking code. So I would say:
```javascript
if (!a) return;
if (!b) return;
if (!c) return;
do_something_important();
```
is easier to read (and thus preferable) than:
```javascript
if (a) {
  if (b) {
    if (c) {
      do_something_important();
    }
  }
}
```

---

<sup>**[modules/tactical/src/data_manager.ts, line 40 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4FmGGSaRbOwdWI-yD)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L40)):</sup>
Done.

---

<sup>**[modules/tactical/src/data_manager.ts, line 64 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BOll5IGPdRn5r7Iu)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L64)):</sup>
Done.

---

<sup>**[modules/tactical/src/link.ts, line 21 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BVzOjzOGRVA81bGh)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L21)):</sup>
Yep. I couldn't get it to work before, but it seems happy now.

---

<sup>**[modules/tactical/src/link.ts, line 29 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0Y1OJ-EEgIgj-JJ9u-r1-29)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L29)):</sup>
Gone.

---

<sup>**[modules/tactical/src/link.ts, line 30 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4GBHB9yTm9MQzem8b)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L30)):</sup>
Done.

---

<sup>**[modules/tactical/src/link.ts, line 39 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BdTBOyFaAC4YMjO4)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L39)):</sup>
Done.

---

<sup>**[modules/tactical/src/link.ts, line 45 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4Bi9tenQogPFbX0j6)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L45)):</sup>
Done.

---

<sup>**[modules/tactical/src/link.ts, line 51 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt4BjjGtyqHZMt1iZIS)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/link.ts#L51)):</sup>
Usually it's sufficient to comment once and add "apply throughout". :)

---

<sup>**[modules/tactical/test/link_spec.ts, line 48 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt0g74wNvEPHOhBo0lj)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/test/link_spec.ts#L48)):</sup>
We should.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/tactical/24)

**Comment by ttowncompiled** [comment]
Review status: 0 of 6 files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

<sup>**[modules/tactical/src/data_manager.ts, line 33 \[r1\]](https://reviewable.io:443/reviews/angular/tactical/24#-Jt49l7S5ofumsZLteyt)** ([raw file](https://github.com/angular/tactical/blob/8509203a0e04f2a9f425cef9e3e42505b95b8484/modules/tactical/src/data_manager.ts#L33)):</sup>
In hindsight, my comment earlier comes off as really rude. I apologize for that.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/tactical/24)

**Pull request merged**